File: frontend/src/types/venta.ts
// types/venta.ts

export enum EstadoVenta {
  PENDIENTE_PAGO = 'PENDIENTE_PAGO',
  COMPLETADA = 'COMPLETADA',
  CANCELADA = 'CANCELADA'
}

export enum MetodoPago {
  EFECTIVO = 'EFECTIVO',
  TRANSFERENCIA = 'TRANSFERENCIA',
  QR = 'QR'
}

export enum ModalidadEntrega {
  RETIRO = 'RETIRO',
  DOMICILIO = 'DOMICILIO',
  TRANSPORTADORA = 'TRANSPORTADORA'
}

export enum EstadoEntrega {
  PENDIENTE = 'PENDIENTE',
  ENTREGADO = 'ENTREGADO'
}

export interface ItemVentaRequest {
  idProducto: number;
  cantidad: number;
  precioUnitarioConDescuento?: number;
  descuentoPorcentaje?: number;
}

/**
 * Request para crear venta.
 *
 * Dos modos:
 *  1. Cliente registrado: enviar idCliente.
 *  2. Venta de mostrador: enviar nombreClienteInvitado (y opcionalmente el
 *     teléfono). El backend crea un cliente tipo INVITADO con esos datos, en
 *     lugar de guardar el nombre suelto dentro de la venta como antes.
 *
 * El metodoPago viaja acá porque alimenta el primer pago de la venta, no un
 * campo de la venta: una venta admite varios cobros con métodos distintos.
 */
export interface VentaRequest {
  // MODO 1: Cliente registrado
  idCliente?: number;

  // MODO 2: Venta de mostrador
  nombreClienteInvitado?: string;
  telefonoClienteInvitado?: string;

  // Datos comunes
  metodoPago: MetodoPago;
  referenciaPago?: string;
  items: ItemVentaRequest[];

  // Monto efectivamente cobrado. Si no viene, el backend asume pago total.
  montoPagado?: number;

  // Entrega. Si no viene, el backend asume RETIRO.
  modalidadEntrega?: ModalidadEntrega;
  direccionDestino?: string; // obligatorio para DOMICILIO y TRANSPORTADORA
  ciudad?: string; // obligatorio para DOMICILIO y TRANSPORTADORA
  transportadora?: string; // obligatorio para TRANSPORTADORA
  guiaRemision?: string; // obligatorio para TRANSPORTADORA
}

export interface DetalleVenta {
  id: number;
  idProducto: number;
  nombreProducto: string;
  skuProducto?: string;
  cantidad: number;
  precioUnitario: number;
  subtotal: number;
}

export interface Pago {
  id: number;
  idVenta?: number;
  monto: number;
  metodoPago: MetodoPago;
  referencia?: string;
  observacion?: string;
  estado: string;
  fechaPago: string;
  urlComprobante?: string | null;
  sinRespaldo?: boolean;
}

export interface Venta {
  id: number;
  idCliente?: number;
  nombreCliente: string;
  telefonoCliente?: string;
  fechaVenta: string;

  // Importes: los cuatro que guarda la tabla, no solo el total
  subtotal: number;
  descuento?: number;
  montoTotal: number;
  saldoPendiente?: number;

  estado: EstadoVenta;
  requiereEnvio?: boolean;

  modalidadEntrega?: ModalidadEntrega;
  estadoEntrega?: EstadoEntrega;
  direccionDestino?: string;
  ciudad?: string;
  transportadora?: string;
  guiaRemision?: string;

  /**
   * Método de pago mostrado en los listados. Ya no es un campo de la venta:
   * el backend lo deriva de los pagos y devuelve el método cuando hay uno
   * solo, o "VARIOS" cuando hay más de uno. Por eso es string y no MetodoPago.
   */
  metodoPago?: string;

  idUsuario: number;
  nombreUsuario: string;
  detalles: DetalleVenta[];
  pagos: Pago[];
  fechaActualizacion: string;
  esClienteRegistrado: boolean;

  /** Al menos un pago QR/transferencia no tiene foto de comprobante todavía. */
  tienePagosSinRespaldo?: boolean;
}

export interface VentaEstadisticas {
  totalVentas: number;
  ventasCompletadas: number;
  ventasPendientes: number;
  ventasCanceladas: number;
  montoTotal: number;
  ventasDelDia: number;
  montoDelDia: number;
}

File: app/lib/models/venta.dart
// Modelos de venta: cliente, importes, entrega y sus pagos. Salen de
// GET /api/ventas y GET /api/ventas/{id} (VentaResponse en el backend).

enum EstadoVenta { completada, pendientePago, cancelada }

EstadoVenta estadoVentaDesdeApi(String valor) {
  switch (valor) {
    case 'COMPLETADA':
      return EstadoVenta.completada;
    case 'CANCELADA':
      return EstadoVenta.cancelada;
    case 'PENDIENTE_PAGO':
    default:
      return EstadoVenta.pendientePago;
  }
}

enum EstadoEntrega { pendiente, entregado }

EstadoEntrega estadoEntregaDesdeApi(String valor) {
  return valor == 'ENTREGADO' ? EstadoEntrega.entregado : EstadoEntrega.pendiente;
}

enum ModalidadEntrega { retiro, domicilio, transportadora }

ModalidadEntrega modalidadEntregaDesdeApi(String? valor) {
  switch (valor) {
    case 'DOMICILIO':
      return ModalidadEntrega.domicilio;
    case 'TRANSPORTADORA':
      return ModalidadEntrega.transportadora;
    case 'RETIRO':
    default:
      return ModalidadEntrega.retiro;
  }
}

enum MetodoPago { efectivo, transferencia, qr }

extension MetodoPagoApi on MetodoPago {
  /// Valor que espera el backend (MetodoPago en Java).
  String get valorApi {
    switch (this) {
      case MetodoPago.efectivo:
        return 'EFECTIVO';
      case MetodoPago.transferencia:
        return 'TRANSFERENCIA';
      case MetodoPago.qr:
        return 'QR';
    }
  }

  String get etiqueta {
    switch (this) {
      case MetodoPago.efectivo:
        return 'Efectivo';
      case MetodoPago.transferencia:
        return 'Transferencia';
      case MetodoPago.qr:
        return 'QR';
    }
  }
}

MetodoPago? metodoPagoDesdeApi(String? valor) {
  switch (valor) {
    case 'EFECTIVO':
      return MetodoPago.efectivo;
    case 'TRANSFERENCIA':
      return MetodoPago.transferencia;
    case 'QR':
      return MetodoPago.qr;
    default:
      return null;
  }
}

class DetalleVenta {
  const DetalleVenta({
    required this.nombreProducto,
    required this.skuProducto,
    required this.cantidad,
    required this.precioUnitario,
    required this.subtotal,
  });

  final String nombreProducto;
  final String? skuProducto;
  final int cantidad;
  final double precioUnitario;
  final double subtotal;

  factory DetalleVenta.desdeApi(Map<String, dynamic> json) {
    return DetalleVenta(
      nombreProducto: json['nombreProducto'] as String? ?? '',
      skuProducto: json['skuProducto'] as String?,
      cantidad: (json['cantidad'] as num?)?.toInt() ?? 0,
      precioUnitario: (json['precioUnitario'] as num?)?.toDouble() ?? 0.0,
      subtotal: (json['subtotal'] as num?)?.toDouble() ?? 0.0,
    );
  }
}

class Pago {
  const Pago({
    required this.id,
    required this.monto,
    required this.metodoPago,
    required this.fechaPago,
    required this.referencia,
    required this.nombreUsuario,
  });

  final int id;
  final double monto;
  final MetodoPago? metodoPago;
  final DateTime? fechaPago;
  final String? referencia;
  final String? nombreUsuario;

  factory Pago.desdeApi(Map<String, dynamic> json) {
    return Pago(
      id: json['id'] as int,
      monto: (json['monto'] as num?)?.toDouble() ?? 0.0,
      metodoPago: metodoPagoDesdeApi(json['metodoPago'] as String?),
      fechaPago: json['fechaPago'] != null ? DateTime.tryParse(json['fechaPago'] as String) : null,
      referencia: json['referencia'] as String?,
      nombreUsuario: json['nombreUsuario'] as String?,
    );
  }
}

class Venta {
  const Venta({
    required this.id,
    required this.nombreCliente,
    required this.telefonoCliente,
    required this.fechaVenta,
    required this.montoTotal,
    required this.saldoPendiente,
    required this.estado,
    required this.modalidadEntrega,
    required this.estadoEntrega,
    required this.direccionDestino,
    required this.ciudad,
    required this.transportadora,
    required this.guiaRemision,
    required this.detalles,
    required this.pagos,
  });

  final int id;
  final String nombreCliente;
  final String? telefonoCliente;
  final DateTime? fechaVenta;
  final double montoTotal;
  final double saldoPendiente;
  final EstadoVenta estado;
  final ModalidadEntrega modalidadEntrega;
  final EstadoEntrega estadoEntrega;
  final String? direccionDestino;
  final String? ciudad;
  final String? transportadora;
  final String? guiaRemision;
  final List<DetalleVenta> detalles;
  final List<Pago> pagos;

  bool get tieneSaldoPendiente => saldoPendiente > 0;

  factory Venta.desdeApi(Map<String, dynamic> json) {
    final detallesJson = json['detalles'] as List<dynamic>? ?? const <dynamic>[];
    final pagosJson = json['pagos'] as List<dynamic>? ?? const <dynamic>[];

    return Venta(
      id: json['id'] as int,
      nombreCliente: json['nombreCliente'] as String? ?? 'Cliente no especificado',
      telefonoCliente: json['telefonoCliente'] as String?,
      fechaVenta: json['fechaVenta'] != null ? DateTime.tryParse(json['fechaVenta'] as String) : null,
      montoTotal: (json['montoTotal'] as num?)?.toDouble() ?? 0.0,
      saldoPendiente: (json['saldoPendiente'] as num?)?.toDouble() ?? 0.0,
      estado: estadoVentaDesdeApi(json['estado'] as String? ?? 'PENDIENTE_PAGO'),
      modalidadEntrega: modalidadEntregaDesdeApi(json['modalidadEntrega'] as String?),
      estadoEntrega: estadoEntregaDesdeApi(json['estadoEntrega'] as String? ?? 'PENDIENTE'),
      direccionDestino: json['direccionDestino'] as String?,
      ciudad: json['ciudad'] as String?,
      transportadora: json['transportadora'] as String?,
      guiaRemision: json['guiaRemision'] as String?,
      detalles: detallesJson.map((d) => DetalleVenta.desdeApi(d as Map<String, dynamic>)).toList(),
      pagos: pagosJson.map((p) => Pago.desdeApi(p as Map<String, dynamic>)).toList(),
    );
  }
}
